Speed up tests by running the biggest ones first

kwiberg opened the pull request:
And default to one thread per CPU.

kwiberg commented:
Updated the pull request. (Patch 2/3: Now store just the latest time for each test in a single file. Patch 3/3: Eliminated the named function.)

pbos commented on gtest-parallel:
```diff
@@ -82,6 +87,45 @@ class RawFormat:
+class TestTimes(object):
+  def __init__(self, save_file):
+    "Create new object seeded with saved test times from the given file."
+    self.__old_times = {}  # (test binary, test name) -> runtime in seconds
```
Do you need both old_times and new_times? get_test_time is only called before tests are running, right?

Also, imo remove the __ prefixes, unless you can point me to somewhere saying they're Google style.

kwiberg replied:
I think you're right regarding old_times and new_times. I'll combine them.

Prefixing with __ isn't about style, it's how member variables are made private in Python. See e.g. https://docs.python.org/2/tutorial/classes.html#private-variables-and-class-local-references.

pbos commented on gtest-parallel:
```diff
@@ -82,6 +87,45 @@ class RawFormat:
+
+    try:
+      with gzip.GzipFile(save_file, "rb") as f:
+        self.__old_times = cPickle.load(f)
```
Is there any chance that you can validate that old_times is now a (test binary, test name) -> runtime in seconds? Or at least (string, string) -> float so we don't get any runtime errors if we change file format.

kwiberg replied:
That's a good idea. Will do.

pbos commented on gtest-parallel:
```diff
@@ -175,11 +236,12 @@ def run_job((command, job_id, test)):
   code = sub.wait()
-  runtime_ms = int(1000 * (time.time() - begin))
-  logger.log(str(job_id) + ': EXIT ' + str(code) + ' ' + str(runtime_ms))
+  runtime = time.time() - begin
```
runtime_ms so we can be consistent, any sub-ms precision doesn't matter either way

kwiberg committed:
Speed up tests by running the biggest ones first

Towards the end of a test run, there's typically a period when only
one or a few cores are busy with the last of the tests---the ones that
take a very long time. By sorting the tests according to their
expected runtime and starting the longest-running ones first, we
minimize this effect.

This patch records the runtime in ~/.gtest-parallel-times once all
tests have finished; on subsequent runs, this data is used to start
the longest-running tests first. Speedup varies depending on which
tests are run, but the usual seems to be about 10%, and I haven't
observed a slowdown for any set of tests (though admittedly, my sample
is very modest).

kwiberg committed:
Catch OSError when listing tests

I got an uncaught exception here when passing in the name of a
non-existing test binary. Now, the error message contains the name of
the missing file instead of a stack trace.

kwiberg committed:
Default to one thread per CPU, instead of 16 regardless of the number of CPUs

This should be a better default for systems with few (or very many)
cores. Also, since we began starting the longest-running tests first,
there's no longer any need for extra parallelism just to get those
tests to start earlier.
